### ISSUE TITLE
Research Keycloak "Operating Modes"

### ISSUE THREAD

**Issue by hkeeler** [cfpb/hmda-platform-auth#78]
Per Keycloak [Choosing an Operating Mode](http://www.keycloak.org/docs/2.5/server_installation_guide/topics/operating-mode.html):
>Before deploying Keycloak in a production environment you need to decide which type of operating mode you are going to use. Will you run Keycloak within a cluster? Do you want a centralized way to manage your server configurations? Your choice of operating mode effects how you configure databases, configure caching and even how you boot the server.

We need to understand these modes (Standalone vs. Standalone Clustered vs. Domain Clustered), and figure out the pros and cons of each, and how they could be deployed to our various environments.

**Comment by hkeeler** [comment]
I made a lot of progress on this issue.  Here's the current state:

1. I have done a lot of research on Keycloak's various _operating modes_.  [Standalone Clustered](https://keycloak.gitbooks.io/server-installation-and-configuration/content/topics/operating-mode/standalone-ha.html) mode seems to make the most sense for both our for our Docker and Compose/DCOS environments.  At a high level, each Keycloak instance is considered a peer, and data is sync'd between each node via [JGroups](http://jgroups.org/). 
1. I have a working prototype with multiple Keycloak instances, and an Apache reverse proxy sitting in front of them, using mod_proxy_balancer.  In this setup, we see the following:
    1. Traffic is equally balanced between the two Keycloak instances.
    1. When a given instance is taken offline, all traffic is routed to the still-up server, with only a slight pause noticeable from a UX perspective. 👍 
    1. When a downed instance is brought back online, traffic stops being routed to _either_ Keycloak instance.  This seems to be a mod_proxy_balancer + Compose (DNS cache?) type issue.  I have not been able to confirm. 👎 
1. The next step is to try this setup in our dev cluster.  There are many, many ways the reverse proxy, Keycloak, and JGroups can be setup, and I won't know for sure which makes the most sense or us until I deploy and test there.  I also believe many of the issues I've seen with Compose will not be an issue in that environment, but I won't know for sure until I try.

**Comment by hkeeler** [comment]
We've spent a good of deal time researching Keycloak running in clustered mode.  We have been able to launch Keycloak in clustered mode, but we have not been fix of some of the JGroups-related errors and failures that occur when this service runs for an extended period of time.  For now, we are discontinuing research on this, and will stick to standalone mode.  If time permits, we will pick it back up in the future.
